lib/restrict_admin_by_ip_middleware.rb

```diff
@@ -0,0 +1,30 @@
+class RestrictAdminByIpMiddleware
+  def initialize(app, allowed_ips)
+    @app = app
+    @allowed_ips = allowed_ips.map { |ip| IPAddr.new(ip) }
+  end
+
+  def call(env)
+    if restricted?(env)
+      [403, {"Content-Type" => "text/plain"}, ["Forbidden"]]
+    else
+      @app.call(env)
+    end
+  end
+
+  private
+
+  def restricted?(env)
+    req = Rack::Request.new(env)
+
+    restricted_route?(req) && !allowed_ip?(req)
+  end
+
+  def restricted_route?(req)
+    req.path == "/admin" || req.path.start_with?("/admin/")
```

Comment:
Is it not enough to do `.start_with?("/admin")` here?

Reply:
What if we have a route `/admin-report` or something (a concieted example, I know) which we don't want to block? Since this is hidden in the middleware, and all environments other than production let anyone in, we might well not notice until it went live.